### PR TITLE
Stackdriver monitoring AlertPolicy resource

### PR DESCRIPTION
**Pull request by chrisst**
Adding the AlertPolicy resource
All stack driver resources are condition upon https://github.com/GoogleCloudPlatform/magic-modules/pull/497 getting accepted.

-----------------------------------------------------------------
# [all]
## [terraform]
Stackdriver monitoring AlertPolicy resource

### CODE REVIEW

File: products/monitoring/terraform.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+--- !ruby/object:Provider::Terraform::Config
+overrides: !ruby/object:Provider::ResourceOverrides
+  AlertPolicy: !ruby/object:Provider::Terraform::ResourceOverride
+    exclude: false
+    id_format: "{{name}}"
+    examples: |
```
Comment (rileykarson) [inline products/monitoring/terraform.yaml]: Complete drive-by but wanna use a new-style example :)

#423 #446 #485 for context

File: products/monitoring/api.yaml [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+           name: mutateTime
+           description: |
+             When the change occurred.
+         - !ruby/object:Api::Type::String
+           name: mutatedBy
+           description: |
+             The email address of the user making the change.
+      - !ruby/object:Api::Type::Boolean
+        name: enabled
+        # TODO fix MM to accept default boolean values
```
[the remark this answers is not in the document]
Comment (chrisst) [reply products/monitoring/api.yaml]: It was more difficult then I thought it would be. It's some [meta logic in type.rb](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/api/type.rb#L76) but because ruby doesn't actually have a normal 'boolean' class and instead has native classes of TrueClass and FalseClass it will take a little bit of work. I'll give it a shot though and see if I can get it in.

[the remark this answers is not in the document]
Comment (chrisst) [reply products/monitoring/api.yaml]: ❓ should this be `send_empty_value: true` so that somebody can send enabled false?

File: products/monitoring/api.yaml [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+      guides:
+        'Official Documentation':
+          'https://cloud.google.com/monitoring/docs/'
+      api: 'https://cloud.google.com/monitoring/api/ref_v3/rest/'
+    properties:
+      - !ruby/object:Api::Type::String
+        name: parentName
+        description: |
+          The name of the group's parent, if it has one. The format is
+          "projects/{project_id_or_number}/groups/{group_id}". For
```
[the remark this answers is not in the document]
Comment (chrisst) [reply products/monitoring/api.yaml]: At this point I haven't done any work/testing on group it's mostly slurped up from hawk's commits so I'm going to wait to release it until I can put time towards it. Since I don't plan on releasing Group as part of this PR I'll add as a TODO for when I get Group finished.

File: products/monitoring/terraform.yaml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                }
+              ]
+            }
+          }
+        ]
+      }
+      ```
+    properties:
+      name: !ruby/object:Provider::Terraform::PropertyOverride
+        custom_expand: 'templates/terraform/custom_expand/monitoring_alert_policy_name.erb'
```
[the remark this answers is not in the document]
Comment (chrisst) [reply products/monitoring/terraform.yaml]: I'm not totally sure tbh - @SirGitsalot added this logic so maybe he has a better answer?

File: products/monitoring/terraform.yaml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        custom_flatten: 'templates/terraform/custom_flatten/name_from_self_link.erb'
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
+      post_create: templates/terraform/post_create/alert_policy.erb
+
+  Group: !ruby/object:Provider::Terraform::ResourceOverride
+    exclude: true
```
[the remark this answers is not in the document]
Comment (chrisst) [reply products/monitoring/terraform.yaml]: It's because it isn't ready. I added details [in the commit](https://github.com/GoogleCloudPlatform/magic-modules/pull/503/commits/9e47223eec4d85bde6fff74146c45d2ee830320a) but a comment makes a lot more sense in hindsight :P 

TBH the only reason I left this in here rather then stripping it or commenting it out was so that I could preserve the git blame history. However I don't feel super strongly about it. I'm happy to go with whichever you prefer (add comment+exclude or rip it out and add back later).

**Comment by chrisst** [comment]
@danawillow last few commits should clean up the name/id logic and address the review comments. 
I have 1 question regarding preference for handling the Group resource.

File: products/monitoring/api.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (danawillow) [reply products/monitoring/api.yaml]: probably, but the easiest way to find out is going to be to try it out and see what happens

File: products/monitoring/terraform.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (danawillow) [reply products/monitoring/terraform.yaml]: Ok! I'm also fine either way. Let's do comment+exclude then (but put the exclude in api.yaml)

File: templates/terraform/post_create/set_computed_name.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// `name` is autogenerated from the api so needs to be set post-create
+name, ok := res["name"]
+if !ok {
+  return fmt.Errorf("Create response didn't contain critical fields. Create may not have succeeded.")
+}
+
+d.Set("name", name.(string))
+d.SetId(name.(string))
```
Comment (danawillow) [inline templates/terraform/post_create/set_computed_name.erb]: i thought we were going to set the id to the display name so that we didn't have to have custom code around it

File: templates/terraform/tests/resource_monitoring_alert_policy_test.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+		}
+
+		project, err := getTestProject(rs.Primary, config)
+		if err != nil {
+			return err
+		}
+
+		name := rs.Primary.Attributes["name"]
+
+		url := fmt.Sprintf("hthttps://monitoring.googleapis.com/v3/projects/%s/alertPolicies/%s", project, name)
```
Comment (danawillow) [inline templates/terraform/tests/resource_monitoring_alert_policy_test.go]: typo: https

File: products/monitoring/api.yaml [hunk 8]
```diff
@@ -0,0 +1,754 @@
...
+                  contain restrictions on resource type,
+                  resource labels, and metric labels. This
+                  field may not exceed 2048 Unicode characters
+                  in length.
+           - !ruby/object:Api::Type::String
+             name: name
+             description: |
+               The unique resource name for this condition.
+               Its syntax is: projects/[PROJECT_ID]/alertPolicies/[POLICY_ID]/con
+               ditions/[CONDITION_ID] [CONDITION_ID] is assigned
```
Comment (danawillow) [inline products/monitoring/api.yaml]: this should be on the same line as the previous one so it doesn't end up with a space in between "con" and "ditions" in the generated markdown

File: templates/terraform/post_create/set_computed_name.erb (hunk 6 above)
Comment (chrisst) [reply templates/terraform/post_create/set_computed_name.erb]: You are correct and I forgot to update you after I ran into a couple different problems: I'm stuck using custom code in order to set `name` as a post-create anyway. When importing the resource we need to have a way to construct the GET url and right now the import logic is using the id. So I think for now the id has to contain enough information to construct a the resource url without heavy modification to the import logic.

Also I was rushing on Friday and didn't realize I had broke my import tests because the import logic also has an assumption that `/`'s have significance in import_id and will parse them out. So I'm going to just go back to commenting as much of this as possible into hawk's old strategy :(

**Comment by chrisst** [comment]
I added another block of fields to the alertPolicy in https://github.com/chrisst/magic-modules/pull/1 but figured this PR was getting messy enough that they could be reviewed in their own PR. I can either open another PR against master or you can review the current PR which is against this branch

File: products/monitoring/terraform.yaml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+              ]
+            }
+          }
+        ]
+      }
+      ```
+    properties:
+    # The upstream API expects name to be `projects/[PROJECT_ID]/alertPolicies/[POLICY_ID]` but
+    # we are storing just the {POLICY_ID} as the id/name in terraform state. This is because 
+    # import logic constructs a url from the id as isn't currently tolerant of id's that are
```
Comment (danawillow) [inline products/monitoring/terraform.yaml]: That's not true though, right? The import fn itself doesn't call get, only the read does, and it does it based on the self_link_url set in api.yaml, which in this case would have been {{name}}. The import fn would have just set name in state (you should be able to override the list of allowed import values to make it only be {{name}} or write a custom import function). The id itself is never read anywhere.

I want to make sure we get this right now, since I think this sort of thing is going to pop up a lot in the future (https://cloud.google.com/apis/design/standard_fields) and I want to make sure we maintain the ability for people to correctly interpolate based on fields that are in the API. If another resource needed to refer to an AlertPolicy, I assume it would do so based on its full name (unfortunately stackdriver doesn't have any examples on this, so I don't know for sure), which means we want to store that full name _somewhere_, and if we can have it just match the API that would be really nice.

Comment (chrisst) [reply products/monitoring/terraform.yaml]: I might have been muddling things, but I'll see if I can explain a bit more: At this point we'd need to rewrite the regex generation (more details below) to use `import_format`. I did forget that the Read method is called with all the context still set on the resource so I was thinking that the id was being used but you're right that it's not actually being used.

re:regex - Changing the `import_format` to `{{name}} with the way the regex is currently built will translate into regex of `(?P<name>[^/]+)` where it's getting tripped up by the fact there is a slash in the name and not matching the whole id.

I hadn't thought about just using a custom import, so that would probably be a much cleaner option then what I have going so far. Probably the 'right' way would be to fix the regex, but I'm pretty sure it would take changing the id_format input significantly or introducing something like `id_regex` so I was hesitant to do that without discussion strategy in advance.
